docs/dev_101_services.md

```diff
@@ -197,3 +197,62 @@ If you need more control over the service call, you can also pass an async funct
 async def custom_set_sleep_timer(entity, service_call):
     await entity.set_sleep_timer(service_call.data['sleep_time'])
 ```
+
+## Response Data
+
+Services may optionally respond to a service call with data for powering more advanced automations. The use of response data for services is meant for
+cases that do not fit the Home Assistant state. For example, a response stream of repeated objects, or API would be better queried on demand or filtered such as a search.
+
+Example code:
+
+```python
+import datetime
+import voluptuous as vol
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, ServiceCall, ServiceResponse
+from homeassistant.helpers import config_validation as cv, entity_platform, service
+
+
+SEARCH_ITEMS_SERVICE_NAME = "search_items"
+SEARCH_ITEMS_SCHEMA = vol.Schema({
+    vol.Required("start"): datetime.datetime,
+    vol.Required("end"): datetime.datetime,
+})
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the platform."""
+
+    async def search_items(call: ServiceCall) -> ServiceResponse:
+        """Search in the date range and return the matching items."""
+        if not call.return_response:
+            raise ValueError("Request did not ask for response data")
+        items = await my_client.search(call.data["start"], call.data["end"])
+        return {
+            "items": [
+                {
+                    "summary": item["summary"],
+                    "description": item["description"],
+                } for item in items
+            ],
+        }
+
+      hass.services.async_register(
+          DOMAIN,
+          SEARCH_ITEMS_SERVICE_NAME,
+          search_items,
+          schema=SEARCH_ITEMS_SCHEMA,
+          supports_response=True,
+      )
+```
+
+There are some additional implementation standards:
```

Comment:
I would move this before the code sample, and separate the must haves from optional items. This way the reader has already a context when they skim over the code example, and they know which parts they are required to implement.

Reply:
I restructured this based on your feedback.